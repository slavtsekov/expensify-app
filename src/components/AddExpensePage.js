import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { requestAddExpense } from "../actions/expenses";

export class AddExpensePage extends React.Component {
    onSubmit = (data) => {
        this.props.startAddExpense(data);
        this.props.history.push("/");
    };
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm
                        onSubmit={this.onSubmit}
                    />
                </div>
            </div>            
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (data) => { dispatch(requestAddExpense(data)); }
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);