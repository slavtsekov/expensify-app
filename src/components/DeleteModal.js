import React from "react";
import Modal from "react-modal";

const DeleteModal = (props) => (
    <Modal
        isOpen={props.isOpen}
        onRequestClose={props.handleDecline}
        contentLabel="Selected Option"
        closeTimeoutMS={200}
        className="modal"
    >
        <h3 className="modal__title">Delete</h3>
        <p className="modal__body">Are you sure you want to delete the selected expense?</p>
        <button className="button modal__confirm" onClick={props.handleAccept}>OK</button>
        <button className="button button--secondary" onClick={props.handleDecline}>Cancel</button>
    </Modal>
);

export { DeleteModal as default };