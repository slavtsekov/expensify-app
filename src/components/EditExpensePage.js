import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense2, startRemoveExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component {
    editExpense = (updated) => { 
        this.props.startEditExpense(this.props.expense.id, updated);
        this.props.history.push("/");
    };
    removeExpense = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push("/");
    };
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm
                        expense={this.props.expense}
                        onSubmit={this.editExpense} 
                    />
                    <button 
                        className="button button--secondary" 
                        onClick={this.removeExpense}>
                        Remove Expense
                    </button>
                </div>
            </div>            
        );
    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
    startEditExpense: (id, updated) => { dispatch(startEditExpense2(id, updated)); },
    startRemoveExpense: (data) => { dispatch(startRemoveExpense(data)); }
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);