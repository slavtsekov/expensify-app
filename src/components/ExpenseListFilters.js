import React from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from "../actions/filters";

export class ExpenseListFilters extends React.Component {
    state = {
        focusedDateInput: null
    };
    changeTextFilter = (e) => {
        this.props.setTextFilter(e.target.value);
    };
    changeSortBy = (e) => {
        if (e.target.value === "amount") {
            this.props.sortByAmount();
        } else {
            this.props.sortByDate();
        }
    };
    changeDateFilter = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    changeFocusedDateInput = (focusedDateInput) => {
        this.setState({ focusedDateInput })
    };
    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input 
                            type="text" 
                            value={this.props.filters.text}
                            onChange={this.changeTextFilter} 
                        />                    
                    </div>
                    <div className="input-group__item">
                        <select
                            value={this.props.filters.sortBy}
                            onChange={this.changeSortBy}>
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            startDateId="start_date_id"
                            endDate={this.props.filters.endDate}
                            endDateId="end_date_id"
                            onDatesChange={this.changeDateFilter}
                            focusedInput={this.state.focusedDateInput}
                            onFocusChange={this.changeFocusedDateInput}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                            showClearDates={true}
                        />
                    </div>
                </div>
            </div>
        );  
    }  
}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (value) => { dispatch(setTextFilter(value)); },
    sortByAmount: () => { dispatch(sortByAmount()); },
    sortByDate: () => { dispatch(sortByDate()); },
    setStartDate: (startDate) => { dispatch(setStartDate(startDate)); },
    setEndDate: (endDate) => { dispatch(setEndDate(endDate)); }
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);