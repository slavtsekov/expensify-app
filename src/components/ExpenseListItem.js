import React from "react";
import { Link } from "react-router-dom";

const ExpenseListItem = (props) => {
    const { id, description, amount, createdAt } = props.data;
    return (
        <div>
            <Link to={`/edit/${id}`}>{description}</Link> - {amount} - {createdAt}
        </div>
    );
};

export default ExpenseListItem;