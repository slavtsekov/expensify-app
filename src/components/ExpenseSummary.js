import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import numeral from "numeral";
import selectExpenses from "../selectors/expenses.js";
import selectExpensesTotal from "../selectors/expenses-total.js";

export const ExpenseSummary = ({ expenseCount, expensesTotal, hiddenExpensesCount }) => {
    const visibleExpensesWord = expenseCount !== 1 ? "expenses" : "expense";
    const formattedExpensesTotal = numeral(expensesTotal / 100).format("$0,0.00");
    const hiddenExpensesWord = hiddenExpensesCount !== 1 ? "expenses" : "expense";
    
    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">
                    Viewing <span>{expenseCount}</span> {visibleExpensesWord} totalling <span>{formattedExpensesTotal}</span>.
                </h1>
                {hiddenExpensesCount > 0 && (
                    <h2 className="page-header__title subtitle">
                        Not showing <span>{hiddenExpensesCount}</span> {hiddenExpensesWord} because of filters.
                    </h2>
                )}
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add Expense</Link>
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);

    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses),
        hiddenExpensesCount: state.expenses.length - visibleExpenses.length
    };
};

export default connect(mapStateToProps)(ExpenseSummary);