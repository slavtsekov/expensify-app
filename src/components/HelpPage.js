import React from "react";

const HelpPage = () => (
    <div>
        This is my help component.
    </div>
);

export { HelpPage as default };