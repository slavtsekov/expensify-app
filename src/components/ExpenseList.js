import React from "react";
import { connect } from "react-redux";
import selectExpenses from "../selectors/expenses";

import ExpenseListItem from "./ExpenseListItem";

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length !== 0 ? (
                props.expenses.map((expense) => (
                    <ExpenseListItem key={expense.id} data={expense} />
                ))
            ) : (
                <p>No expenses</p>
            )
        }
    </div>
);

const mapStateToProps = (state) => ({
    expenses: selectExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);