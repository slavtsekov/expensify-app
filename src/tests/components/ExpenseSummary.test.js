import React from "react";
import { shallow } from "enzyme";
import { ExpenseSummary } from "../../components/ExpenseSummary";

test("should render ExpenseSummary with one expense", () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={1} expensesTotal={344} />);
    expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseSummary with multiple expenses", () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={45} expensesTotal={56785} />);
    expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseSummary with hidden expense", () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={3} expensesTotal={23} hiddenExpensesCount={1} />);
    expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseSummary with multiple hidden expenses", () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={3} expensesTotal={23} hiddenExpensesCount={5} />);
    expect(wrapper).toMatchSnapshot();
});