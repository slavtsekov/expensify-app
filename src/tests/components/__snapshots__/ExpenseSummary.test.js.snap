// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary with hidden expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        3
      </span>
       
      expenses
       totalling 
      <span>
        $0.23
      </span>
      .
    </h1>
    <h2
      className="page-header__title subtitle"
    >
      Not showing 
      <span>
        1
      </span>
       
      expense
       because of filters.
    </h2>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseSummary with multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        45
      </span>
       
      expenses
       totalling 
      <span>
        $567.85
      </span>
      .
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseSummary with multiple hidden expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        3
      </span>
       
      expenses
       totalling 
      <span>
        $0.23
      </span>
      .
    </h1>
    <h2
      className="page-header__title subtitle"
    >
      Not showing 
      <span>
        5
      </span>
       
      expenses
       because of filters.
    </h2>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseSummary with one expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $3.44
      </span>
      .
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
